src/client/keycode.h

```diff
@@ -4,62 +4,86 @@
 
 #pragma once
 
-#include "exceptions.h"
 #include "irrlichttypes.h"
+#include "cmake_config.h" // USE_SDL2
 #include <Keycodes.h>
 #include <IEventReceiver.h>
 #include <string>
+#include <unordered_map>
+#include <variant>
 
-class UnknownKeycode : public BaseException
-{
-public:
-	UnknownKeycode(const char *s) :
-		BaseException(s) {};
-};
-
-/* A key press, consisting of either an Irrlicht keycode
-   or an actual char */
-
+/* A key press, consisting of a scancode or a keycode */
 class KeyPress
 {
 public:
 	KeyPress() = default;
 
-	KeyPress(const char *name);
+	KeyPress(std::string_view name);
 
-	KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character = false);
+	KeyPress(const irr::SEvent::SKeyInput &in);
 
-	bool operator==(const KeyPress &o) const
+	std::string sym(const bool force_scancode = false) const;
+	std::string name() const;
+
+	irr::EKEY_CODE getKeycode() const;
+	wchar_t getKeychar() const;
+
+	//TODO: drop this once we fully switch to SDL
+	u32 getScancode() const
 	{
-		return (Char > 0 && Char == o.Char) || (valid_kcode(Key) && Key == o.Key);
+#if USE_SDL2
+		if (auto pv = std::get_if<u32>(&scancode))
+			return *pv;
+#endif
+		return 0;
 	}
 
-	const char *sym() const;
-	const char *name() const;
+	bool operator==(const KeyPress &o) const
+	{
+#if USE_SDL2
+		return scancode == o.scancode;
+#else
+		return (Char > 0 && Char == o.Char) || (Keycode::isValid(Key) && Key == o.Key);
+#endif
+	}
 
-protected:
-	static bool valid_kcode(irr::EKEY_CODE k)
+	operator bool() const
 	{
-		return k > 0 && k < irr::KEY_KEY_CODES_COUNT;
+#if USE_SDL2
+		return std::holds_alternative<irr::EKEY_CODE>(scancode) ?
+			Keycode::isValid(std::get<irr::EKEY_CODE>(scancode)) :
+			std::get<u32>(scancode) != 0;
+#else
+		return Char > 0 || Keycode::isValid(Key);
+#endif
 	}
 
+	static const KeyPress &getSpecialKey(const std::string &name);
+
+private:
+#if USE_SDL2
+	bool loadFromScancode(std::string_view name);
+	std::string formatScancode() const;
+
+	std::variant<u32, irr::EKEY_CODE> scancode = irr::KEY_UNKNOWN;
+#else
 	irr::EKEY_CODE Key = irr::KEY_KEY_CODES_COUNT;
 	wchar_t Char = L'\0';
-	std::string m_name = "";
+#endif
+
+	static std::unordered_map<std::string, KeyPress> specialKeyCache;
 };
 
 // Global defines for convenience
-
-extern const KeyPress EscapeKey;
-
-extern const KeyPress LMBKey;
-extern const KeyPress MMBKey; // Middle Mouse Button
-extern const KeyPress RMBKey;
+// This implementation defers creation of the objects to make sure that the
+// IrrlichtDevice is initialized.
+#define EscapeKey KeyPress::getSpecialKey("KEY_ESCAPE")
```

Comment:
I must have missed it but which information does `KeyPress` hold that is not available without an SDL device?
the scancode?
and it wouldn't work if it contained `irr::KEY_ESCAPE`?